### ISSUE TITLE
improve the sematic of lr_scheduler

### ISSUE THREAD

**Issue by mli** [dmlc/mxnet#625]
the current lr_scheduler is based on epoch, it is not flexible, and it has problem with the distributed training, where the optimizer runs on the server nodes. a server node neither can get the current epoch number, nor there is a clear definition of epoch for async sgd. 

i suggest to turn to a `batch` based lr scheduler implemented in python/mxnet, while provide an epoch like interface to users, say

```
new_lr = .8 * old_lr for every .5 epoch
```

we then set a correct batch value for lr_scheduler.

but for a correct implementation, `batch` is not a good term. it means reading a data batch, while not necessary a weight updating:

1. we can sum the gradient for several batches and then apply one weight updating. often useful if the model is huge, while gpu mem is little, especially on the distributed setting.
2. on distributed training, assume n machines and b batches, there are n*b weight updatings for async_sgd while only b updatings for sync_sgd

so i suggest to pass `num_updates` to the lr_scheduler. but there there are still two issues

1. we may update layer 1 (conv layer) for each batch, while layer 2 (big fullc layer) for every 2 batches.
2. assume there are two workers, on async_sgd, the possible updating order on the server node could be
```
layer_1_worker_1, layer_2_worker_2, layer_1_worker_2, layer_2_worker_1
``` 
to simplify the thing, we pass "the maximal number of updates applied to a weight" to the lr_scheduler, for the above example, the values are

```
1, 1, 2, 2
```

**Comment by pluskid** [comment]
Can we put learning rate in kvstore and synchronize it?

**Comment by mli** [comment]
currently it only syncs once when call `kvstore.set_optimizer(optm)`. syncing later is nontrivial in a full async updating, some nodes may fast while others may slow, especially, any worker may die at any moment if running on the ec2 spot instance

**Comment by piiswrong** [comment]
This will result in different learning rate for each layer right?

**Comment by piiswrong** [comment]
With async sgd, each thread push their own grad update on their own weight, I think they should all increment num_updates by one.

**Comment by mli** [comment]
no if set num_update = "the maximal number of updates applied to a weight", so every layer share the same upper bound. we can support defining a scheduler for each layer, but it seems over-engineering.

**Comment by jermainewang** [comment]
Which scenarios are you talking about? SGD or ASGD? I think in no matter
which way the update and the adjustment of learning rate should happen on
server side.

   - In distributed SGD, the minibatch will be carved into smaller pieces
   and the update will be performed after gathering all worker’s gradient.
   - In distributed ASGD, after each minibatch, the update will happen.

lr_scheduler should not be aware of either case, so the code could remain
the same. I think you could still keep num_batch since in SGD case, the
update is still performed after one minibatch. In either case, one *logical*
minibatch still equals to one update.

I dont understand the second problem you mentioned. Why would two layers
update in different frequency? Is this still valid SGD?
​

On Wed, Nov 18, 2015 at 6:48 PM, Mu Li <notifications@github.com> wrote:

> no if set num_update = "the maximal number of updates applied to a
> weight", so every layer share the same upper bound. we can support defining
> a scheduler for each layer, but it seems over-engineering.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dmlc/mxnet/issues/625#issuecomment-157901108>.
>



-- 
Minjie Wang
*New York University | Computer Science*
715 Broadway, New York, NY, 10009

**Comment by mli** [comment]
one example is the alexnet, it has a few conv layers and three big fullc layers. in order to reduce the communication overhead but still enjoy a small batch size, we can sync & update the conv layers for each batch, while sync the fullc layers for every 10 batches.

**Comment by jermainewang** [comment]
Oh i see. This the variable batch size trick in Alex paper. Do we have plan
to implement that?

On Wed, Nov 18, 2015 at 10:08 PM, Mu Li <notifications@github.com> wrote:

> one example is the alexnet, it has a few conv layers and three big fullc
> layers. in order to reduce the communication overhead but still enjoy a
> small batch size, we can sync & update the conv layers for each batch,
> while sync the fullc layers for every 10 batches.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dmlc/mxnet/issues/625#issuecomment-157934970>.
>



-- 
Minjie Wang
*New York University | Computer Science*
715 Broadway, New York, NY, 10009

**Comment by mli** [comment]
probably not in a near future except that we going to train a distributed vgg or other large models
